src/Control/Effect/Carrier.hs

```diff
@@ -1,79 +1,151 @@
 --
 --   1. Be functorial in their last two arguments, and
 --   2. Support threading effects in higher-order positions through using the carrier’s suspended state.
 --
--- All first-order effects (those without recursive occurrences of @m@) admit a default definition
--- of 'handle'. The @-XDeriveAnyClass@ extension allows derivation of both 'HFunctor' and 'Effect':
---
--- @
---   data State s (m :: * -> *) k
---     = Get (s -> k)
---     | Put s k
---       deriving (Functor, HFunctor, Effect)
--- @
+-- All first-order effects (those without existential occurrences of @m@) admit a default definition of 'handle' provided a 'Generic1' instance is available for the effect.
 class HFunctor sig => Effect sig where
   -- | Handle any effects in a signature by threading the carrier’s state all the way through to the continuation.
-  handle :: Functor f
+  handle :: (Functor f, Monad m)
          => f ()
          -> (forall x . f (m x) -> n (f x))
-         -> sig m (m a)
-         -> sig n (n (f a))
-
-  default handle :: (Functor f, Coercible (sig m (n (f a))) (sig n (n (f a))))
+         -> sig m a
+         -> sig n (f a)
+  default handle :: (Functor f, Monad m, Generic1 (sig m), Generic1 (sig n), GEffect m n (Rep1 (sig m)) (Rep1 (sig n)))
                  => f ()
                  -> (forall x . f (m x) -> n (f x))
-                 -> sig m (m a)
-                 -> sig n (n (f a))
-  handle state handler = coerce . fmap' (handler . (<$ state))
+                 -> sig m a
+                 -> sig n (f a)
+  handle state handler = to1 . ghandle state handler . from1
   {-# INLINE handle #-}
 
 
-
 -- | The class of carriers (results) for algebras (effect handlers) over signatures (effects), whose actions are given by the 'eff' method.
 class (HFunctor sig, Monad m) => Carrier sig m | m -> sig where
   -- | Construct a value in the carrier for an effect signature (typically a sum of a handled effect and any remaining effects).
-  eff :: sig m (m a) -> m a
+  eff :: sig m a -> m a
 
 -- | Apply a handler specified as a natural transformation to both higher-order and continuation positions within an 'HFunctor'.
-handlePure :: HFunctor sig => (forall x . f x -> g x) -> sig f (f a) -> sig g (g a)
-handlePure handler = hmap handler . fmap' handler
+handlePure :: (HFunctor sig, Functor f) => (forall x . f x -> g x) -> sig f a -> sig g a
+handlePure = hmap
 {-# INLINE handlePure #-}
+{-# DEPRECATED handlePure "handlePure has been subsumed by hmap." #-}
 
 -- | Thread a 'Coercible' carrier through an 'HFunctor'.
 --
 --   This is applicable whenever @f@ is 'Coercible' to @g@, e.g. simple @newtype@s.
-handleCoercible :: (HFunctor sig, Coercible f g) => sig f (f a) -> sig g (g a)
-handleCoercible = handlePure coerce
+handleCoercible :: (HFunctor sig, Functor f, Coercible f g) => sig f a -> sig g a
+handleCoercible = hmap coerce
 {-# INLINE handleCoercible #-}
+
+
+-- | Generic implementation of 'HFunctor'.
+class GHFunctor m m' rep rep' where
```

Comment:
Do we need to INLINE the definitions here? Can we check the generated Core/add to the inspection tests to make sure we’re not missing these opportunities?